server/rest.go

```diff
@@ -788,6 +788,35 @@ func (s *Server) AddressHolders(c echo.Context) error {
 		krcTokenInfo, _ := s.getKRCTokenInfo(ctx, holders[i].ContractAddress)
 		if krcTokenInfo != nil {
 			holders[i].Logo = krcTokenInfo.Logo
+			// get holder balance from RPC for rechecking
+			if strings.EqualFold(krcTokenInfo.TokenType, cfg.SMCTypeKRC20) {
+				smcABIStr, err := s.cacheClient.SMCAbi(ctx, cfg.SMCTypePrefix+cfg.SMCTypeKRC20)
+				if err != nil {
+					s.logger.Warn("Cannot get KRC20 token ABI", zap.Error(err), zap.String("smcAddress", holders[i].ContractAddress))
+					continue
+				}
+				smcABI, err := s.decodeSMCABIFromBase64(ctx, smcABIStr, holders[i].ContractAddress)
+				if err != nil {
+					s.logger.Warn("Cannot decode KRC20 token ABI", zap.Error(err), zap.String("smcAddress", holders[i].ContractAddress))
+					continue
+				}
+				balance, err := s.kaiClient.GetKRC20BalanceByAddress(ctx, smcABI, common.HexToAddress(holders[i].ContractAddress), common.HexToAddress(holders[i].HolderAddress))
+				if err != nil {
+					s.logger.Warn("Cannot get KRC20 balance of address", zap.Error(err), zap.String("holderAddress", holders[i].HolderAddress),
+						zap.String("smcAddress", holders[i].ContractAddress))
+					continue
+				}
+				s.logger.Info("RPC balance vs db balance", zap.String("RPC", balance.String()), zap.String("DB", holders[i].BalanceString))
+				if !strings.EqualFold(balance.String(), holders[i].BalanceString) {
+					// update correct balance to database and return to client
+					holders[i].BalanceString = balance.String()
+					holders[i].BalanceFloat = s.calculateKRC20BalanceFloat(balance, krcTokenInfo.Decimals)
+					err = s.dbClient.UpdateHolders(ctx, []*types.TokenHolder{holders[i]})
+					if err != nil {
+						s.logger.Warn("Cannot update KRC20 holder with new balance", zap.Error(err), zap.Any("holder", holders[i]))
+					}
```

Comment:
Could you explain your idea and intention when you update the balance here?

Reply:
I update the correct balance for statistical and database sorting reason.

Reply:
what happened in case the user got transferred token without access this API ?

Reply:
In this case, the token balance might be wrong because grabber will make a contract call to a random RPC endpoint in list, we are not sure if the state of that RPC node has applied the new block yet. It's the reason why issue https://github.com/kardiachain/kardia-explorer-backend/issues/104 happened.

Reply:
if so, the root cause is we should retrieve data in grabber with 1 node, instead of fetch user balance from the network for each user request?